nebula-logger/core/main/logger-engine/classes/Logger.cls

```diff
@@ -2686,36 +2635,80 @@ global with sharing class Logger {
      * @return   The value currently set as the current transaction's scenario
      */
     global static String getScenario() {
-        return transactionScenario;
+        return currentEntryScenario;
     }
 
     /**
-     * @description Sets the scenario name for the current transaction - this is stored in `LogEntryEvent__e.Scenario__c`
-     *              and `Log__c.Scenario__c`, and can be used to filter & group logs
+     * @description Sets the current scenario, which can be used to identify modules or groupings of  for the current transaction
      * @param  scenario The name to use for the current transaction's scenario
      */
+    @SuppressWarnings('PMD.NcssMethodCount')
     global static void setScenario(String scenario) {
-        if (USE_FIRST_SPECIFIED_SCENARIO == true && String.isNotBlank(transactionScenario) == true) {
-            return;
+        if (USE_FIRST_SPECIFIED_SCENARIO_AS_TRANSACTION_SCENARIO == false || String.isBlank(transactionScenario) == true) {
+            transactionScenario = scenario;
         }
 
-        transactionScenario = scenario;
+        if (String.isNotBlank(scenario) == true && scenario == currentEntryScenario) {
+            return;
+        }
 
-        LogScenarioRule__mdt matchingScenarioRule;
-        for (LogScenarioRule__mdt scenarioRule : LogScenarioRule__mdt.getAll().values()) {
-            if (scenarioRule.IsEnabled__c == true && scenarioRule.Scenario__c == scenario) {
-                matchingScenarioRule = scenarioRule;
-                break;
+        currentEntryScenario = scenario;
+        orderedScenarios.add(scenario);
+        LoggerScenarioRule__mdt matchingScenarioRule = LoggerScenarioRule.getInstance(scenario);
+        if (matchingScenarioRule != null) {
+            LoggerSettings__c userSettings = getUserSettings();
+            if (String.isNotBlank(matchingScenarioRule.IsLoggerEnabled__c) == true) {
```

Comment:
I think that we did chat about this earlier, but can you refresh my memory on the thought process behind implementing these as picklists? Also, does it make sense to have one global value set for them if they're all going to be true/false?

Reply:
I want all of these fields (that override a field on `LoggerSettings__c`) to be optional overrides - that way, if you're creating your own `LoggerScenarioRule__mdt`, you can choose to only override a subset of `LoggerSettings__c` fields. You may still want `LoggerSettings__c` to control _most_ of the settings. But since `Boolean` fields can't be null in Salesforce, the easiest (although a bit gross) way I've found to make it optional is to make it an optional picklist field, and use a `null` check in the code to see if the field should be overridden.

And I just double checked in my scratch org - for some mysterious reason, you can't use a global value set for picklists on a custom metadata type object. It seems like a feature gap on the platform unfortunately.

![image](https://user-images.githubusercontent.com/1267157/191793445-2105eb36-60b8-476f-8975-a587d883e7cf.png)

Reply:
Ah yes, I remember now. I knew we had talked about this! Thanks for clarifying.